dask_planner/src/sql/optimizer/join_reorder.rs

```diff
@@ -0,0 +1,385 @@
+                _extract_inner_joins(&join.left, rels, conds);
+                _extract_inner_joins(&join.right, rels, conds);
+
+                for (l, r) in &join.on {
+                    conds.insert((l.clone(), r.clone()));
+                }
+
+                // Need to save this info somewhere
+                let join_filter = join.filter.as_ref().unwrap();
+            } */
+            _ => {
+                if find_join(plan).is_some() {
+                    for x in plan.inputs() {
+                        _extract_inner_joins(x, rels, conds);
+                    }
+                } else {
+                    // Leaf node
+                    rels.push(plan.clone())
+                }
+            }
+        }
+    }
+
+    let mut rels = vec![];
+    let mut conds = HashSet::new();
+    _extract_inner_joins(plan, &mut rels, &mut conds);
+    (rels, conds)
+}
+
+/// Find first (top-level) join in plan
+fn find_join(plan: &LogicalPlan) -> Option<Join> {
+    match plan {
+        LogicalPlan::Join(join) => Some(join.clone()),
+        other => {
+            if other.inputs().is_empty() {
+                None
+            } else {
+                for input in &other.inputs() {
+                    if let Some(join) = find_join(input) {
+                        return Some(join);
+                    }
+                }
+                None
+            }
+        }
+    }
+}
+
+fn get_unfiltered_dimensions(dims: &[Relation]) -> Vec<Relation> {
+    dims.iter().filter(|t| !t.has_filter()).cloned().collect()
+}
+
+fn get_filtered_dimensions(dims: &[Relation]) -> Vec<Relation> {
+    dims.iter().filter(|t| t.has_filter()).cloned().collect()
+}
+
+fn build_join_tree(
+    fact: &LogicalPlan,
+    dims: &[LogicalPlan],
+    conds: &mut HashSet<(Column, Column)>,
+) -> Result<LogicalPlan> {
+    let mut b = LogicalPlanBuilder::from(fact.clone());
+    for dim in dims {
+        // Find join keys between the fact and this dim
+        let mut join_keys = vec![];
+        for (l, r) in conds.iter() {
+            if (b.schema().index_of_column(l).is_ok() && dim.schema().index_of_column(r).is_ok())
+                || b.schema().index_of_column(r).is_ok() && dim.schema().index_of_column(l).is_ok()
+            {
+                join_keys.push((l.clone(), r.clone()));
+            }
+        }
+        if !join_keys.is_empty() {
+            let left_keys: Vec<Column> = join_keys.iter().map(|(l, _r)| l.clone()).collect();
+            let right_keys: Vec<Column> = join_keys.iter().map(|(_l, r)| r.clone()).collect();
+
+            for key in join_keys {
+                conds.remove(&key);
+            }
+
+            /* FIXME: Build join with join_keys when needed
+            self.join(
+                right: LogicalPlan,
+                join_type: JoinType,
+                join_keys: (Vec<impl Into<Column>>, Vec<impl Into<Column>>),
+                filter: Option<Expr>,
```

Comment:
Probably the biggest improvement we need is to support join filters correctly. This should allow us to run query 72, which is expected to have the largest performance gain with join reordering. Other queries affected include queries 75 and 93.